docs/building.md

```diff
@@ -0,0 +1,176 @@
+# Cloning and building
+
+Fobnail project consists of 3 main repositories with code:
+
+- [Fobnail firmware](https://github.com/fobnail/fobnail) (Rust)
+- [Platform Owner](https://github.com/fobnail/fobnail-platform-owner) (C)
+- [Attester](https://github.com/fobnail/fobnail-attester) (C)
+
+There is also [Fobnail SDK](https://github.com/fobnail/fobnail-sdk) that
+simplifies building of firmware by using Docker container, but only
+[one script](https://github.com/fobnail/fobnail-sdk/blob/main/run-fobnail-sdk.sh)
+from this repository is actually needed.
+
+Following steps were successfully performed on Ubuntu 22.04.1 LTS.
+
+## Fobnail Token firmware
+
+Start by installing prerequisites and adding current user to `docker` group:
+
+```shell
+sudo apt -y install git docker.io
+sudo usermod -aG docker $USER
+```
+
+Log out of the desktop session then log in again - this is required after making
+the change to user groups. You may confirm it by running `groups`.
+
+Clone the repository along with its submodules:
+
+```shell
+git clone https://github.com/fobnail/fobnail --recurse-submodules
+```
+
+Build process for Token firmware is by far the most complicated one. For that
+reason Fobnail SDK is strongly suggested. To use it, simply download script
+`run-fobnail-sdk.sh` either by cloning whole repository or just this one script
+and save it to a directory in your `PATH` environment variable:
+
+```shell
+sudo wget -O /usr/local/bin/run-fobnail-sdk.sh \
+  https://raw.githubusercontent.com/fobnail/fobnail-sdk/main/run-fobnail-sdk.sh
+sudo chmod +x /usr/local/bin/run-fobnail-sdk.sh
+```
+
+With the script installed, one can prepare for the build process itself.
+Firmware can be build for a physical Fobnail Token or simulated on PC, with
+different preparatory steps required. In both cases, first execution of
+`build.sh` downloads Docker container image, which can be both time- and
+bandwidth-consuming.
+
+#### Environment variables common for both targets
+
+Fobnail Token firmware is configured with environment variables passed to
+`build.sh`.
+
+Note that due to the way Docker mounts directories all files and directories
+pointed to by following variables must be located somewhere in `fobnail`
+directory.
+
+- `FOBNAIL_PO_ROOT` - **required** option, must point to valid PEM or DER file
+  with Platform Owner's root certificate. See [this document](/keys_and_certificates/#platform-owner-certificate-chain)
```

Comment:
Another thing that is confusing, at least to me, is if I should generate just PO Root CA, or maybe all CAs mentioned in section linked (Platform Owner certificate chain -> [Maximum certificate chain length](https://fobnail.3mdeb.com/keys_and_certificates/#informational-checks))?

In section [Maximum certificate chain length](https://fobnail.3mdeb.com/keys_and_certificates/#informational-checks) we talking about Root CA, intermediate CA and final CA, but we not explaining how to generate final one. We just say about Root CA and intermediate CA. It is explained that:

> During Fobnail provisioning, Platform Owner generates one additional certificate for Token, but this final certificate doesn't allow for use as a CA.

I don't know if I'm doing provisioning while trying to accomplish "Cloning and building" from "Development Process". 

Let's clarify in simplest possible way what is expected from user to "Cloning and building" of "Development Process". Any advanced considerations can be added as additional notes - those information should not be mixed with basic content because it cause documentation very hard to follow.

Reply:
So, for building only root certificate is required. Other certificates won't be needed until Token provisioning. I will add a sentence telling so here.

> In section [Maximum certificate chain length](https://fobnail.3mdeb.com/keys_and_certificates/#informational-checks) we talking about Root CA, intermediate CA and final CA, but we not explaining how to generate final one

From that section:

> Maximal supported certificate chain consists of 3 CA certificates: root CA (flashed to Fobnail's memory), intermediate CA and final CA that is used to issue certificates for keys generated by Fobnail. Intermediate CA is optional, root and issuing CA aren't.

Isn't that together with [Intermediate/issuing CA](https://fobnail.3mdeb.com/keys_and_certificates/#intermediateissuing-ca) clear enough?

I don't think there are any `advanced considerations`, this whole documentation tries to describe simplest possible, configurable, reasonably secure use of Fobnail. Security is never easy and always relative. We can't just pack everything in one script and satisfy the needs of all.

Reply:
> So, for building only root certificate is required. Other certificates won't be needed until Token provisioning. I will add a sentence telling so here.

Great. Please link to commit so I can resolve that part of conversation.

> Isn't that together with [Intermediate/issuing CA](https://fobnail.3mdeb.com/keys_and_certificates/#intermediateissuing-ca) clear enough?

I tried to read in the way is suggested by documentation. I click link to understand `FOBNAIL_PO_ROOT` and I'm getting lecture about certificate chains. What I would expect from section "Cloning and building" is exactly that cloning and building not deep dive into design decisions mad in Fobnail. As a new user I have plenty of time for that after I decide project is interesting to me.

> I don't think there are any advanced considerations, this whole documentation tries to describe simplest possible, configurable, reasonably secure use of Fobnail.

We can have our own opinions. For you it is not advanced and perfectly ok to lead one in presented way. I tried to use documentation as new user/developer and expressed my opinion. It is completely fine we have different opinions but when writing documentation it is important for whom we writing if we want this project to achieve some traction from community.

From your comment it sounds you wan to provide usable reasonably secure configuration, but IIUC project is far from something for every day use. Please correct me if I'm wrong.

We should also consider [Diataxis](https://diataxis.fr/) as it was endorsed by many projects and some 3mdeb team members.

> Security is never easy and always relative. We can't just pack everything in one script and satisfy the needs of all.

Agree, but our goal should be to simplify that for end users and potential developers. Part of the project success is wrapping complexity. You can find that concept in many popular solutions, which we use every day, which behind the scene leveraging advanced security concept 99.99% population has not idea about.

Reply:
> If only thing the user see is few pages of build and deploy instructions after which the only visible result is some blinking LEDs, then it is hard to decide whether "magic" that in the end result in blinking lights is interesting or not. 

We will definitely work on making it more on explaining things in simplest possible words, but now we have what we have. If there are no ideas on improving those demos then it has to stay like that. There are plenty of complex projects which describe things in user approachable way, but it needs better writing and marketing skills from our side. Let's leave this topic for now.

> I guess this can be improved by linking to Architecture instead of or in addition to Detailed description on the main page.

Please do.

> Apart from strange race conditions coming probably from systemd that sometimes require few tries, I'd say that demo from incoming blog post is ready to use.

Do we have tracking issue for those race conditions?

> I agree, and that is partially what I tried to do with these changes and those from previous phase. On the other hand, this sometimes is against DRY, instructions for building FOBNAIL_PO_ROOT instead of full chain at once is a good example of such case.

This always can happen, but let perfect not be enemy of good. If breaking DRY wins users I'm ok with that.

> Few months ago the documentation was just a bunch of unrelated how-to's, most of which were outdated. Now the project in my opinion is in stage where proper, targeted documentation makes sense, however preparing such documentation is a significant effort. Usually it also requires additional people, ignorant (in a good sense of the word) to the project, because developers which worked on it from the start take a lot of things for granted, which I learned when we decided to hand demo preparation to a different team.

Agree, I'm ignorant and I'm willing to make review here and find problems. I don't know if I ever would have time to improve documentation personally, but we can also looks for support in other places when we will identify what have to be improved.

> Again, it's all about compromises that have to be made. By making it easier to use Fobnail in one specific scenario, it gets harder to make it configurable for other use cases. Maybe my assumptions about the project are wrong, but I think that Fobnail (as a whole) is for people that are somewhat aware of security issues they are trying to avoid, either on hardware they themselves are using (home PC) or are responsible for, but which are used by other people (office PCs, kiosks or other terminals). I'd rather make it hard to provision, easy and secure to use, because I assume that people responsible for provisioning are more technologically advanced than end users.

This is correct assumption. I guess it is based on your judgement of Fobnail maturity.  In longer run we need both users (I have no idea/rough idea how it works in the background, but it seem to work, give me value and I will better using this open-source security technology) and developers (I can clone, build, compile and hack whatever I need). There is also small group where users and developers overlap, but too many of our projects/products are targeted at developers, we should try to change our thinking. Very good post about [attitude of some coreboot developers](https://www.phoronix.com/forums/forum/phoronix/latest-phoronix-articles/1356230-old-amd-cpu-motherboard-support-removed-from-open-source-coreboot?p=1356254#post1356254) - please note number of likes, user friendliness is important and targeting documentation at specific group helps achieve better experience.

Reply:
> > I guess this can be improved by linking to Architecture instead of or in addition to Detailed description on the main page.
> 
> Please do.

Done.

Reply:
Last thing:
- [ ] Do we have a tracking issue for those race conditions?
- [ ] `Done.` -  can you link me to a particular commit? In such a way, I would solve conversations faster.

Reply:
Race conditions are already fixed, thanks to Maciej.

> Done.

https://github.com/fobnail/docs/pull/47/commits/b924a999c0dedefbafc8ed64f6f7e969722ca27d